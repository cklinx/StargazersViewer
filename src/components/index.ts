export * from './background/Background';
export * from './button/Button';
export * from './info-text-layout/InfoTextLayout';
