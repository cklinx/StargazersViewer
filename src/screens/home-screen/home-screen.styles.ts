import {Platform, StyleSheet} from 'react-native';
import {SCREEN_WIDTH} from '@stargazers/utils/dimensions';
import {Fonts} from '@stargazers/theme';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 0,
    width: '100%',
    alignSelf: 'center',
    alignItems: 'center',
    justifyContent: 'center',
  },
  panel: {
    width: '85%',
    maxWidth: 400,
    padding: 20,
    paddingTop: 15,
    height: 310,
    borderWidth: 1,
    borderRadius: 20,
    borderColor: '#35485b',
    borderBottomWidth: 0,
    shadowColor: '#000000',
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.4,
    shadowRadius: 2,
    elevation: 1,
    position: 'absolute',
    transform: [{translateY: 180}],
    zIndex: 2000,
    ...Platform.select({
      ios: {},
      android: {
        marginTop: -10,
        shadowOffset: {width: 0, height: 2},
        elevation: 15,
      },
    }),
  },
  panelSync: {
    width: 150,
    height: 150,
    padding: 20,
    paddingTop: 15,
    borderWidth: 1,
    borderRadius: 20,
    borderColor: '#FFF',
    borderBottomWidth: 0,
    shadowColor: '#000000',
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 0.9,
    shadowRadius: 6,
    elevation: 3,
    position: 'absolute',
    transform: [{translateY: 800}],
    zIndex: 1200,
    ...Platform.select({
      ios: {},
      android: {
        marginTop: -10,
        shadowOffset: {width: 0, height: 5},
        elevation: 7,
      },
    }),
  },
  logoContainer: {
    flex: 1,
    position: 'absolute',
    top: 69,
  },
  image: {
    width: 100,
    height: 80,
    marginBottom: 10,
    marginLeft: 53,
    overflow: 'visible',
  },
  overlay: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    backgroundColor: 'rgba(255,255,255,0.12)',
    zIndex: 1000,
  },
  footer: {
    paddingHorizontal: 22,
    bottom: 0,
    fontSize: 12,
    lineHeight: 20,
    position: 'absolute',
    maxWidth: SCREEN_WIDTH,
  },
  footerLinks: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    alignContent: 'space-between',
    paddingBottom: 10,
    fontFamily: Fonts.helveticaMedium,
    fontSize: 14,
  },
  autocompleteTextInput: {
    borderRadius: 25,
    backgroundColor: '#383b42',
    color: '#fff',
    paddingLeft: 18,
  },
  rightButtonsContainerStyle: {
    right: 15,
    height: 30,
    alignSelf: 'center',
  },
  inputContainerStyle: {
    backgroundColor: '#383b42',
    borderRadius: 25,
    marginTop: 20,
  },
  suggestionContainerStyle: {
    backgroundColor: '#383b42',
  },
  autocompleteContainerStyle: {flexGrow: 1, flexShrink: 1},
  autocompleteItemText: {color: '#fff', padding: 15, fontFamily: Fonts.helveticaMedium},
  continueButton: {
    minWidth: 150,
    alignSelf: 'center',
    borderRadius: 40,
    padding: 10,
  },
  continueButtonText: {
    textAlign: 'center',
    fontFamily: Fonts.helveticaMedium,
    fontSize: 18,
  },
  footerText: {
    fontFamily: Fonts.helveticaMedium,
    fontSize: 13,
    color: '#000',
  },
  emptyListPlaceholder: {
    fontFamily: Fonts.helveticaMedium,
    fontSize: 16,
    color: '#FFF',
    padding: 10,
    textAlign: 'center',
  },
});
