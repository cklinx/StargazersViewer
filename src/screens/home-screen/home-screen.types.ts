import {TAutocompleteDropdownItem} from 'react-native-autocomplete-dropdown';

export interface Autocomplete extends TAutocompleteDropdownItem {}
export interface OwnerSuggestionType extends TAutocompleteDropdownItem {}
