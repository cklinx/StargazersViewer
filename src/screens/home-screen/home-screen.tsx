import React, {useCallback, useEffect, useRef, useState} from 'react';
import {Platform, Animated, Text} from 'react-native';
import useColorScheme from '../../hooks/useColorScheme';
import {OwnerSuggestionType} from '@stargazers/screens/home-screen/home-screen.types';
import Easing from '@stargazers/utils/easing';
import {Colors} from '@stargazers/theme';
import {styles} from './home-screen.styles';
import {AutocompleteDropdown, AutocompleteDropdownRef} from 'react-native-autocomplete-dropdown';
// @ts-ignore
import Feather from 'react-native-vector-icons/Feather';
import {SCREEN_HEIGHT} from '@stargazers/utils/dimensions';
import {getUserRepositories, searchUsers} from '@stargazers/services/user-service/user-service';
import {Background, Button, InfoTextLayout} from '@stargazers/components';
import {HomeScreenProps} from '@stargazers/navigators/navigator.types';

export const HomeScreen: (props: HomeScreenProps) => JSX.Element = ({navigation}) => {

  const colorScheme = useColorScheme();

  const [loading, setLoading] = useState(false);
  const [suggestionsList, setSuggestionsList] = useState<Array<OwnerSuggestionType> | undefined>();
  const [selectedUser, setSelectedUser] = useState<string | null>(null);
  const dropdownController = useRef<AutocompleteDropdownRef | null>(null);
  const searchRef = useRef(null);

  const getSuggestions = useCallback(async (q: string) => {
    const filterToken = q;
    if (typeof q !== 'string' || q.length < 3) {
      setSuggestionsList(undefined);
      return;
    }
    setLoading(true);
    const items = await searchUsers(q);
    const suggestions: OwnerSuggestionType[] = items.data.items
      .filter(item => item.login.includes(filterToken))
      .map(item => ({
        id: item.id.toString(),
        title: item.login,
      }));
    setSuggestionsList(suggestions);
    setLoading(false);
  }, []);

  const onClearPress = useCallback(() => {
    setSelectedUser(null);
    setSuggestionsList(undefined);
  }, []);

  const onOpenSuggestionsList = useCallback((isOpened: boolean) => {}, []);

  // Animations Ref
  const fadeAnim = useRef(new Animated.Value(0)).current;
  const transAnim = useRef(new Animated.Value(180)).current;
  const transAnimFooter = useRef(new Animated.Value(0)).current;

  const setLoadingState = useCallback(() => {
    Animated.timing(transAnimFooter, {
      toValue: 200,
      useNativeDriver: true,
      easing: Easing.ease,
      duration: 500,
    }).start();
    Animated.timing(fadeAnim, {
      toValue: 0,
      useNativeDriver: true,
      easing: Easing.easeInOutCirc,
      duration: 1200,
    }).start();
    Animated.timing(transAnim, {
      toValue: 800,
      useNativeDriver: true,
      easing: Easing.easeInOutCirc,
      duration: 1300,
    }).start();
  }, [fadeAnim, transAnim, transAnimFooter]);

  useEffect(() => {
    const unsubscribe = navigation.addListener('focus', () => {
      console.log('scren focus');
      // Start login animation
      Animated.timing(fadeAnim, {
        toValue: 1,
        useNativeDriver: true,
        easing: Easing.easeInCirc,
        duration: 2200,
      }).start();
      Animated.timing(transAnim, {
        toValue: 0,
        useNativeDriver: true,
        easing: Easing.easeInOutCirc,
        duration: 2300,
      }).start();
    });

    return unsubscribe;
  }, [navigation]);

  const fetchRepositories = () => {
    setLoadingState();
    getUserRepositories(selectedUser!).then(
      response => {
        navigation.navigate('RepositoriesScreen', {
          owner: selectedUser!,
          repositories: response.data,
        });
      },
      () => {
        console.log('error');
      },
    );
  };

  return (
    <>
      <Background>
        <Animated.View
          style={{
            ...styles.footer,
            height: 80,
            transform: [
              {
                translateY: transAnimFooter,
              },
            ],
          }}>
          <Text
            style={{
              ...styles.footerText,
              color: Colors[colorScheme].tint,
            }}>
            {'Stargazers Viewer'}
          </Text>
        </Animated.View>

        <Animated.View
          style={{
            ...styles.panel,
            backgroundColor: Colors[colorScheme].background,
            borderColor: Colors[colorScheme].buttonEnabledColor,
            shadowColor: Colors[colorScheme].secondary,
            opacity: fadeAnim,
            transform: [
              {
                translateY: transAnim,
              },
            ],
          }}>
          <InfoTextLayout
            text={'Type a GitHub user below and select from the list to see his repositories.'}
          />

          <AutocompleteDropdown
            // @ts-ignore
            ref={searchRef}
            controller={controller => {
              dropdownController.current = controller;
            }}
            direction={Platform.select({ios: 'down'})}
            dataSet={suggestionsList}
            onChangeText={getSuggestions}
            onSelectItem={item => {
              item && setSelectedUser(item.title);
            }}
            debounce={600}
            suggestionsListMaxHeight={SCREEN_HEIGHT * 0.4}
            onClear={onClearPress}
            onOpenSuggestionsList={onOpenSuggestionsList}
            loading={loading}
            useFilter={false}
            textInputProps={{
              placeholder: 'Type GitHub user...',
              autoCorrect: false,
              autoCapitalize: 'none',
              style: {
                ...styles.autocompleteTextInput,
              },
            }}
            rightButtonsContainerStyle={styles.rightButtonsContainerStyle}
            inputContainerStyle={styles.inputContainerStyle}
            suggestionsListContainerStyle={{
              ...styles.suggestionContainerStyle,
              backgroundColor: Colors[colorScheme].secondary,
            }}
            containerStyle={styles.autocompleteContainerStyle}
            EmptyResultComponent={
              (
                <Text
                  style={{
                    ...styles.emptyListPlaceholder,
                  }}>
                  No user found
                </Text>
              ) as React.ReactElement<any, any>
            }
            renderItem={item => (
              <Text
                style={{
                  ...styles.autocompleteItemText,
                }}>
                {item.title}
              </Text>
            )}
            ChevronIconComponent={<Feather name="chevron-down" size={20} color="#fff" />}
            ClearIconComponent={<Feather name="x-circle" size={18} color="#fff" />}
            inputHeight={50}
            showChevron={!!selectedUser}
            closeOnBlur={false}
          />

          <Button
            style={{
              ...styles.continueButton,
              backgroundColor:
                selectedUser !== null
                  ? Colors[colorScheme].buttonEnabledBg
                  : Colors[colorScheme].buttonDisabledBg,
              borderColor:
                selectedUser !== null
                  ? Colors[colorScheme].buttonEnabledColor
                  : Colors[colorScheme].buttonDisabledColor,
            }}
            onPress={fetchRepositories}
            disabled={selectedUser === null}>
            <Text
              style={{
                ...styles.continueButtonText,
                color:
                  selectedUser !== null
                    ? Colors[colorScheme].buttonEnabledColor
                    : Colors[colorScheme].buttonDisabledColor,
              }}>
              Continue
            </Text>
          </Button>
        </Animated.View>
      </Background>
    </>
  );
};
