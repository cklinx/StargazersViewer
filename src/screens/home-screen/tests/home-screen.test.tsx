import React from 'react';

import renderer, {act} from 'react-test-renderer';
import {HomeScreen} from '../home-screen';
import {Button} from '@stargazers/components';
import {AutocompleteDropdown} from 'react-native-autocomplete-dropdown';

jest.mock('react-native-vector-icons/Feather', () => 'Feather');
const goBack = jest.fn();
const navigate = jest.fn();
const setOptions = jest.fn();
const addListener = jest.fn();

jest.mock('react-native-autocomplete-dropdown', () => ({
  AutocompleteDropdown: 'AutocompleteDropdown',
}));

describe('HomeScreen', () => {
  it('should render correctly', () => {
    const {toJSON} = renderer.create(
      //@ts-ignore
      <HomeScreen navigation={{goBack, navigate, setOptions, addListener}} />,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('continue button should be disabled', () => {
    const {root} = renderer.create(
      //@ts-ignore
      <HomeScreen navigation={{goBack, navigate, setOptions, addListener}} />,
    );

    const buttonProps = root.findByType(Button).props;
    expect(buttonProps.disabled).toBeTruthy();
  });

  it('continue button should be enabled if user selected', () => {
    const {root} = renderer.create(
      //@ts-ignore
      <HomeScreen navigation={{goBack, navigate, setOptions, addListener}} />,
    );

    const autocompleteProps = root.findByType(AutocompleteDropdown).props;
    act(() => {
      autocompleteProps.onSelectItem({
        title: '4test',
      });
    });
    const buttonProps = root.findByType(Button).props;
    expect(buttonProps.disabled).toBeFalsy();
  });
});
