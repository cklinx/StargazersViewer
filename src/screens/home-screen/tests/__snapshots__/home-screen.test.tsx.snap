// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen should render correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "height": "100%",
      "position": "relative",
      "width": "100%",
    }
  }
>
  <Image
    resizeMode="cover"
    source="test-file-stub"
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "flex": 1,
          "justifyContent": "center",
          "padding": 0,
          "width": "100%",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "bottom": 0,
          "fontSize": 12,
          "height": 80,
          "lineHeight": 20,
          "maxWidth": 750,
          "paddingHorizontal": 22,
          "position": "absolute",
          "transform": Array [
            Object {
              "translateY": 0,
            },
          ],
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#383B42",
            "fontFamily": "HelveticaNeue-Medium",
            "fontSize": 13,
          }
        }
      >
        Stargazers Viewer
      </Text>
    </View>
    <View
      collapsable={false}
      style={
        Object {
          "backgroundColor": "#fff",
          "borderBottomWidth": 0,
          "borderColor": "#FFFFFF",
          "borderRadius": 20,
          "borderWidth": 1,
          "elevation": 1,
          "height": 310,
          "maxWidth": 400,
          "opacity": 0,
          "padding": 20,
          "paddingTop": 15,
          "position": "absolute",
          "shadowColor": "#414757",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.4,
          "shadowRadius": 2,
          "transform": Array [
            Object {
              "translateY": 180,
            },
          ],
          "width": "85%",
          "zIndex": 2000,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#9867F0",
            "borderRadius": 12,
            "flexDirection": "row",
            "marginTop": 12,
            "padding": 16,
            "width": "100%",
          }
        }
      >
        <Feather
          color="#FFFFFF"
          name="alert-circle"
          size={18}
        />
        <Text
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "HelveticaNeue-Medium",
              "fontSize": 16,
              "marginEnd": 16,
              "marginStart": 8,
            }
          }
        >
          Type a GitHub user below and select from the list to see his repositories.
        </Text>
      </View>
      <AutocompleteDropdown
        ChevronIconComponent={
          <Feather
            color="#fff"
            name="chevron-down"
            size={20}
          />
        }
        ClearIconComponent={
          <Feather
            color="#fff"
            name="x-circle"
            size={18}
          />
        }
        EmptyResultComponent={
          <Text
            style={
              Object {
                "color": "#FFF",
                "fontFamily": "HelveticaNeue-Medium",
                "fontSize": 16,
                "padding": 10,
                "textAlign": "center",
              }
            }
          >
            No user found
          </Text>
        }
        closeOnBlur={false}
        containerStyle={
          Object {
            "flexGrow": 1,
            "flexShrink": 1,
          }
        }
        controller={[Function]}
        debounce={600}
        direction="down"
        inputContainerStyle={
          Object {
            "backgroundColor": "#383b42",
            "borderRadius": 25,
            "marginTop": 20,
          }
        }
        inputHeight={50}
        loading={false}
        onChangeText={[Function]}
        onClear={[Function]}
        onOpenSuggestionsList={[Function]}
        onSelectItem={[Function]}
        renderItem={[Function]}
        rightButtonsContainerStyle={
          Object {
            "alignSelf": "center",
            "height": 30,
            "right": 15,
          }
        }
        showChevron={false}
        suggestionsListContainerStyle={
          Object {
            "backgroundColor": "#414757",
          }
        }
        suggestionsListMaxHeight={533.6}
        textInputProps={
          Object {
            "autoCapitalize": "none",
            "autoCorrect": false,
            "placeholder": "Type GitHub user...",
            "style": Object {
              "backgroundColor": "#383b42",
              "borderRadius": 25,
              "color": "#fff",
              "paddingLeft": 18,
            },
          }
        }
        useFilter={false}
      />
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#B4B4B4",
            "borderColor": "#505050",
            "borderRadius": 40,
            "minWidth": 150,
            "padding": 10,
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#505050",
              "fontFamily": "HelveticaNeue-Medium",
              "fontSize": 18,
              "textAlign": "center",
            }
          }
        >
          Continue
        </Text>
      </View>
    </View>
  </View>
</View>
`;
